client/src/sync-filesystem.js

```diff
@@ -6,39 +6,216 @@
 
 var Filer = require('../../lib/filer.js');
 var Shell = require('../../lib/filer-shell.js');
-var Path = Filer.Path;
 var fsUtils = require('../../lib/fs-utils.js');
 var conflict = require('../../lib/conflict.js');
-var resolvePath = require('../../lib/sync-path-resolver.js').resolve;
+var syncModes = require('../../lib/constants.js').syncModes;
+var findPathIndexinArray = require('../../lib/util.js').findPathIndexinArray;
 
 function SyncFileSystem(fs) {
   var self = this;
-  var pathToSync;
-  var modifiedPath;
-
-  // Manage path resolution for sync path
-  Object.defineProperty(self, 'pathToSync', {
-    get: function() { return pathToSync; },
-    set: function(path) {
-      if(path) {
-        pathToSync = resolvePath(pathToSync, path);
-      } else {
-        pathToSync = null;
-      }
+  var root = '/';
+  // Record changes during a downstream sync
+  // Is a parallel array with sourceListCache
+  var recordedPaths = [];
+  var sourceListCache = [];
+  var changesDuringDownstream = [];
+
+  // Expose the root used to sync for the filesystem
+  // Defaults to '/'
+  Object.defineProperties(self, {
+    'root': {
+      get: function() { return root; }
+    },
+    'changesDuringDownstream': {
+      get: function() { return changesDuringDownstream; }
     }
   });
 
-  // Record modifications to the filesystem during a sync
-  Object.defineProperty(fs, 'modifiedPath', {
-    get: function() { return modifiedPath; },
-    set: function(path) {
-      if(path) {
-        modifiedPath = resolvePath(modifiedPath, path);
-      } else {
-        modifiedPath = null;
+  self.record = function(path, sourceList) {
+    recordedPaths.push(path);
+    sourceListCache.push(sourceList);
+  };
+
+  self.stopRecording = function(path) {
+    var indexInRecordedPaths = recordedPaths.indexOf(path);
+    recordedPaths.splice(indexInRecordedPaths, 1);
+    changesDuringDownstream.splice(changesDuringDownstream.indexOf(path), 1);
```

Comment:
Same thing here with changesDuringDownstream.indexOf(path) being -1.  I'd suggest you pull that expression out and create a variable, test the variable, and log the error and maybe bail if you don't find the path when you expect it to be there.